### PR TITLE
tests: be more robust against a new day stepping in

### PR DESCRIPTION
**Pull request by mardy**
Weird things can happen at midnight:

  https://pastebin.ubuntu.com/p/mmQskRXvhM/

The test could fail because it attempts to match the times by their date
component, without taking into account that if the two measurements
happen with the midnight in between, the date component will be
different.

Here we attempt to be a bit more robust, by comparing the given date not
only against the current date, but also against the date we had a few
minutes ago.

### CODE REVIEW

File: tests/main/snapd-state/task.yaml [hunk 1]
```diff
@@ -14,9 +14,19 @@ execute: |
...
     "$TESTSTOOLS"/snapd-state foo 2>&1 | MATCH "snapd-state: no such command: foo"
 
+    function check_date() {
+        local TEST_DATE="$1"
+        local TOLERANCE_SECONDS="1800" # 30 minutes
+
+        CURRENT_DATE="$(date +'%F')"
+        TIMESTAMP_MINUS_TOLERANCE=$(($(date +%s) - TOLERANCE_SECONDS))
+        DATE_MINUS_TOLERANCE="$(date +'%F' --date="@${TIMESTAMP_MINUS_TOLERANCE}")"
+        echo "$TEST_DATE" | MATCH "^(${CURRENT_DATE}|${DATE_MINUS_TOLERANCE})"
```
Comment (stolowski) [inline tests/main/snapd-state/task.yaml:24]: I wonder if we shouldn't change this MATCH-based check (that looks a bit fragile and also hard to understand) to integer comparison instead, i.e. abs of (current date - test timestamp) should be less than tolerance seconds, where dates are converted to seconds with "date +'%s'"?

Comment (mardy) [reply tests/main/snapd-state/task.yaml]: Makes sense indeed, thanks! Done :-)

File: tests/main/snapd-state/task.yaml [hunk 2]
```diff
@@ -14,9 +14,21 @@ execute: |
     "$TESTSTOOLS"/snapd-state --foo 2>&1 | MATCH "snapd-state: no such command: --foo"
     "$TESTSTOOLS"/snapd-state foo 2>&1 | MATCH "snapd-state: no such command: foo"
 
+    function check_date() {
+        local TEST_DATE="$1"
+        local TOLERANCE_SECONDS=1800 # 30 minutes
+
+        local TEST_TIMESTAMP=$(date +'%s' --date="${TEST_DATE}")
```
Comment (stolowski) [inline tests/main/snapd-state/task.yaml:21]: Could please add a comment explaining the problem with this check around midnight, explaining why it's 30 minutes? Do we actually need such high tolerance if we got rid of MATCH?

File: tests/main/snapd-state/task.yaml [hunk 3]
```diff
@@ -14,9 +14,21 @@ execute: |
     "$TESTSTOOLS"/snapd-state --foo 2>&1 | MATCH "snapd-state: no such command: --foo"
     "$TESTSTOOLS"/snapd-state foo 2>&1 | MATCH "snapd-state: no such command: foo"
 
+    function check_date() {
+        local TEST_DATE="$1"
+        local TOLERANCE_SECONDS=1800 # 30 minutes
```
Comment (bboozzoo) [inline tests/main/snapd-state/task.yaml:19]: 5 minutes will probably be sufficient
```suggestion
        local TOLERANCE_SECONDS=300 # 5 minutes
```

**Review by bboozzoo — APPROVED** [review]
LGTM with a little tweak for the tolerance

File: tests/main/snapd-state/task.yaml (hunk 3 above)
Comment (mardy) [reply tests/main/snapd-state/task.yaml:19]: I wish! :-)

If you look at the logs for the failure, the initial timestam was taken at `2021-09-08T23:53:19+00:00` and the failure was detected at `2021-09-09 00:01:57`, that is almost 9 minutes later. Sure, we could reduce the tolerance, but I don't see a benefit in it (and we risk failures on very congested situations).

File: tests/main/snapd-state/task.yaml (hunk 2 above)
Comment (mardy) [reply tests/main/snapd-state/task.yaml]: Done!

**Review by stolowski — APPROVED** [review]
LGTM, +1 when tests are happy. Thank you!

**Pull request merged**